Index: app/src/main/java/com/test/dummyappv3/ui/home/HomeFragment.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- app/src/main/java/com/test/dummyappv3/ui/home/HomeFragment.java	(date 1617101927977)
+++ app/src/main/java/com/test/dummyappv3/ui/home/HomeFragment.java	(date 1617101927977)
@@ -44,7 +44,7 @@
 import java.util.List;
 import com.test.dummyappv3.utils.Utils;
 
-public class HomeFragment extends Fragment {
+public class HomeFragment extends Fragment implements OnQueryTextListener{
 
     private PostViewModel postViewModel;
     public static final String TAG = "HomeFragment";
@@ -276,21 +276,7 @@
         searchView = (SearchView) menu.findItem(R.id.app_bar_search).getActionView();
         searchView.setSearchableInfo(searchManager.getSearchableInfo(getActivity().getComponentName()));
         searchView.setQueryHint(getResources().getString(R.string.searchForPosts));
-        searchView.setOnQueryTextListener(new OnQueryTextListener() {
-            @Override
-            public boolean onQueryTextSubmit(String keyword) {
-                searchState = true;
-                itemArrayList.clear();
-                postViewModel.getItemsBySearch(keyword);
-                adapter.notifyDataSetChanged();
-                return false;
-            }
-
-            @Override
-            public boolean onQueryTextChange(String s) {
-                return false;
-            }
-        });
+        searchView.setOnQueryTextListener(this);
 
         searchView.setOnCloseListener(new SearchView.OnCloseListener() {
             @Override
@@ -361,4 +347,17 @@
         return super.onOptionsItemSelected(item);
     }
 
+    @Override
+    public boolean onQueryTextSubmit(String keyword) {
+        searchState = true;
+        itemArrayList.clear();
+        postViewModel.getItemsBySearch(keyword);
+        adapter.notifyDataSetChanged();
+        return false;
+    }
+
+    @Override
+    public boolean onQueryTextChange(String s) {
+        return false;
+    }
 }
\ No newline at end of file
